keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,336 @@
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP generalizes the existing max-in-flight request handler in the
+apiserver to make more distinctions among requests and provide
+prioritization and fairness among the categories of requests.  An
+outline of the request handling in an apiserver can be found at
+https://speakerdeck.com/sttts/kubernetes-api-codebase-tour?slide=18 .
+
+## Motivation
+
+Today the apiserver has a simple mechanism for protectimg itself
+against CPU and memory overloads: max-in-flight limits for mutating
+and for readonly requests.  Apart from the distinction between
+mutating and readonly, no other distinctions are made among requests;
+consequently there can be undesirable scenarios where one subset of
+the request load crowds out other parts of the request load.  Also,
+the dispatching of these requests against two independent limits is
+not
+[work-conserving](https://en.wikipedia.org/wiki/Work-conserving_scheduler).
+
+### Goals
+
+Following are some bad scenarios that can happen today and which
+should be preventable when this KEP is in place.
+
+- Self-Maintenance crowded out.  Some requests are for system
+  self-maintenance, such as: node heartbeats, kubelet and kube-proxy
+  work on pods, services, secrets, etc involved in the system's
+  self-hosting, and leader elections for system controllers.  In an
+  overload scenario today there is no assurance of priority for these
+  self-maintenance requests.
+
+- Priority Inversions.  In the course of serving request A, there are
```

Comment:
I think this is a future goal, not a current goal.  I would be very happy with a system where is "probably doable later", but I would not make it an objective to start.

Reply:
I think we need to avoid priority inversions from day 1.  It is a well known pathology and there is no reason to think it will not happen in our scenario.

I think that avoiding priority inversions can be approached in a simple way: tell the administrator that the configuration must not give a lower priority to a request that is involved in handling a higher priority request, and that that the case of equal priority should only happen for the highest priority and this priority is not subject to limits.  This is admittedly a fragile solution --- it is only a design rule given to humans.  Later we can envision things like a design rule checker or a more complex design that inherently avoids priority inversions.

Reply:
I don't think it is possible to identify priority inversions statically in any system we could make, because we don't have visibility into what webhooks are doing. Therefore it seems like it solutions have to involve threading contexts through the webhook calls. I think that can and should be done in a second step, and we should solve the problem with documentation in the first pass.

Reply:
That is not to say we couldn't provide a static checker that would be useful, but I think in terms of what is a good use of our time, we will get more benefit from providing a well-commented default set of initial BucketClasses or whatever we will call them.

Reply:
I think we might be saying the same thing with different words.  It is not a goal for this KEP to introduce threading additional information through webhooks.  The idea for this KEP is the simple outline being discussed.  Using that outline, I think it _is_ possible to have some crude avoidance of priority inversions.  This avoidance would be conservative --- it would give high priority to some requests that do not "really" need it; but it would be effective, and I think it would be impractical to not do this in any real deployment of this KEP.  Again, the idea is that the categorization can identify secondary calls from external servers that are serving webhooks in that are invoked in the course of handling a primary call, and we document that the secondary calls be identified and given sufficiently high priority.  Yes, this loses distinctions among primary calls.  The secondary calls have to be given sufficiently high priority for the "worst case".  In other words, the secondary calls have to be given top priority, because we can not rule out the possibility that the secondary call is part of handling a high or top priority primary call.